Fix running `bun test` on multiple `node:test` tests

190n opened the pull request:
### What does this PR do?

Fixes #19200

### How did you verify your code works?

Added to `node-test.test.ts`

robobun commented:
<div><sup>Updated 4:53 PM PT - Apr 28th, 2025</sup></div>

@190n, your commit 9ea7e9397951177ff048d1527d75e6e4826aa50a is building: [`#15769`](https://buildkite.com/bun/bun/builds/15769)
<!-- generated-comment id=oven-sh/bun#19354 -->

jordanebelanger commented:
@190n If this can help you potentially debug, I can still make the tests be skipped under certain circumstances, take this snippet for example:
```typescript
import { test, after } from 'node:test'
import { strict as assert } from 'node:assert'
import { build } from '../helper'

import type { FastifyInstance } from 'fastify'

after(async () => {
  // Cleanup logic if needed
})

test('resource routes', async () => {
  const app: FastifyInstance = await build()

  await test('should create and update a resource', async () => {
    throw new Error('this should fail the test')
  })
})```
When this is ran, the inner test does not get executed (doesn't show up in the console list nor is the thrown error catched/visible)

190n commented:
It looks like we don't intend to support that yet: https://github.com/oven-sh/bun/blob/32789693566b6d9b0fd7583727ebb071cdc1e4eb/src/js/node/test.ts#L146-L148

I think this will be an effort for a different PR. However, I will see if I can figure out why the error isn't being thrown, since that could probably be confusing.

190n commented:
@jordanebelanger the issue was we would only throw the error if you used the context, e.g. using `t.test` in `test("outer", t => t.test("inner", ...))`. I've fixed this and added a test that we emit the error here.

We'll add support for nested tests someday but that's outside the scope of this PR. I think providing an error in this case is still a worthwhile improvement over status quo.

dylan-conway commented on src/bun.js/bindings/BunObject.cpp:
```diff
@@ -748,7 +748,7 @@ JSC_DEFINE_HOST_FUNCTION(functionFileURLToPath, (JSC::JSGlobalObject * globalObj
     listen                                         BunObject_callback_listen                                           DontDelete|Function 1
     udpSocket                                        BunObject_callback_udpSocket                                      DontDelete|Function 1
-    main                                           BunObject_getter_wrap_main                                          DontDelete|PropertyCallback
+    main                                           BunObject_getter_wrap_main                                          DontDelete|CustomAccessor|ReadOnly
```
this is a breaking change making it readonly. does it need to be readonly? could we instead setup a setter?

190n replied:
oh good point, i didn't realize it was previously writable.

readonly was a bit more convenient, but maybe we can have a setter and store an "overridden `Bun.main`" on the VM which the getter checks before figuring out the path, and we can also reset the override when we switch to a different test file? that way if you have

```ts
// 1.test.ts
Bun.main = "overridden";
```

```ts
// 2.test.ts
console.log(Bun.main)
```

it won't remember the override from 1 while running 2

dylan-conway commented on src/bun.js/VirtualMachine.zig:
```diff
@@ -1953,7 +1957,9 @@ extern fn Bun__loadHTMLEntryPoint(global: *JSGlobalObject) *JSInternalPromise;
 pub fn reloadEntryPoint(this: *VirtualMachine, entry_path: []const u8) !*JSInternalPromise {
     this.has_loaded = false;
     this.main = entry_path;
+    this.main_resolved_path = .empty;
```
I think this should deref first, because the previous string impl could only be used in js after this point
```suggestion
    this.main_resolved_path.deref();
    this.main_resolved_path = .empty;
```

190n replied:
oh good point, fixed

dylan-conway commented on src/bun.js/VirtualMachine.zig:
```diff
@@ -2021,7 +2027,9 @@ export fn Bun__VirtualMachine__setOverrideModuleRunMainPromise(vm: *VirtualMachi
 pub fn reloadEntryPointForTestRunner(this: *VirtualMachine, entry_path: []const u8) !*JSInternalPromise {
     this.has_loaded = false;
     this.main = entry_path;
+    this.main_resolved_path = .empty;
```
same here
